### PR TITLE
make kubeadm phases ready for join master

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR implements one of the actions defined by https://github.com/kubernetes/kubeadm/issues/751 (checklist form implementing HA in kubeadm).

With this PR, kubeadm phases implements methods that will be used by the `kubeadm join --master`workflow, and more in detail:
- kubeconfig phase implements a new method for creating kubeconfig required files (nb. with respect to init, the kubelet.conf file should not be created because it will generated by the TLS bootstrap process)
- certs phase implements a new method for checking the pki provided by the users (all the certificates are present, the API server certificate is properly configured)

**Special notes for your reviewer**:
/CC @timothysc @kubernetes/sig-cluster-lifecycle-pr-reviews 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 1]
```diff
@@ -653,6 +653,79 @@ func UsingExternalCA(cfg *kubeadmapi.MasterConfiguration) (bool, error) {
...
+	}
+
+	// front proxy client cert exists
+	if err := validateSignedCert(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.FrontProxyCACertAndKeyBaseName, kubeadmconstants.FrontProxyClientCertAndKeyBaseName, "front-proxy client"}); err != nil {
+		return err
+	}
+
+	// if external etcd, validate if certificates required for connecting API server to etcd exists.
+	// kubeadm does not executes further validations of those certificates
+	if len(cfg.Etcd.Endpoints) > 0 {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/certs.go]: probably cleaner if we return earlier like so:
```php
if len(cfg.Etcd.Endpoints) <= 0 {
	return nil
}

....

return nil		
```

haven't checked if `Endpoints` is an `uint` thus `<= 0` vs `== 0`.

**Comment by fabriziopandini** [comment]
@timothysc removing WIP; implementation is now aligned to the updated KEP 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/CC luxas

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go [hunk 2]
```diff
@@ -59,4 +68,41 @@ func UploadConfiguration(cfg *kubeadmapi.MasterConfiguration, client clientset.I
 			kubeadmconstants.MasterConfigurationConfigMapKey: string(cfgYaml),
 		},
 	})
+	if err != nil {
+		return err
+	}
+
+	// Ensure that the BootstrapDiscoveryClusterRole exists
+	err = apiclient.CreateOrUpdateRole(client, &rbac.Role{
```
Comment (luxas) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: I can't make up my mind whether this is better suited here or in the bootstraptoken phase.
Also, I'd like us to talk about whether it makes sense to do different types of tokens with different privileges for joining a master vs a node.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go [hunk 3]
```diff
@@ -29,6 +30,14 @@ import (
...
 	"k8s.io/kubernetes/cmd/kubeadm/app/util"
 	"k8s.io/kubernetes/cmd/kubeadm/app/util/apiclient"
+	rbachelper "k8s.io/kubernetes/pkg/apis/rbac/v1"
+)
+
+const (
+	// BootstrapDiscoveryClusterRoleName sets the name for the ClusterRole that allows
+	// the bootstrap tokens to access the kubeadm-config ConfigMap during the node bootstrap/discovery
+	// phase for additional master nodes
+	BootstrapDiscoveryClusterRoleName = "kubeadm:bootstrap-discovery-kubeadm-config"
```
Comment (luxas) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: maybe in kubeadmconstants?

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go [hunk 4]
```diff
@@ -73,6 +73,21 @@ func CreateInitKubeConfigFiles(outDir string, cfg *kubeadmapi.MasterConfiguratio
 	)
 }
 
+// CreateJoinMasterKubeConfigFiles will create and write to disk the kubeconfig files required by kubeadm
+// join --master workflow, plus the admin kubeconfig file to be deployed on the new master; the
+// kubelet.conf fil must not be created when joining master nodes because it will be created and signed by
```
Comment (luxas) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: nit: file

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go [hunk 5]
```diff
@@ -73,6 +73,21 @@ func CreateInitKubeConfigFiles(outDir string, cfg *kubeadmapi.MasterConfiguratio
...
+// the kubelet TLS bootstrap process.
+// If any kubeconfig files already exists, it used only if evaluated equal; otherwise an error is returned.
+func CreateJoinMasterKubeConfigFiles(outDir string, cfg *kubeadmapi.MasterConfiguration) error {
+	return createKubeConfigFiles(
+		outDir,
+		cfg,
+		kubeadmconstants.AdminKubeConfigFileName,
```
Comment (luxas) [inline cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: note: this creates `n` different root users to the cluster. Not sure that is desirable from a security standpoint, but it's probably something we should do for a good UX

**Review by luxas — COMMENTED** [review]
kind-a LGTM with one question around security, we'll discuss this first in the KEP you've created :+1:

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: I considered this as well, but currently in pkg `k8s.io/kubernetes/cmd/kubeadm/app/phases/bootstraptoken` there are only things related to the cluster-info configMap.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go (hunk 3 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: This constant is used only here (e.g. similar to constants in pkg `k8s.io/kubernetes/cmd/kubeadm/app/phases/bootstraptoken`). No problem to move it

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go (hunk 4 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: done

File: cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go (hunk 5 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go]: I agree.
As a workaround the user can always copy the admin.conf file from the bootstrap master to the joining master (so there will be only 1 root user, distributed on all the masters)

**Review by luxas — COMMENTED** [review]
Retroactive LGTM

**Pull request merged**